heat/core/tests/test_dndarray.py

```diff
@@ -298,6 +306,62 @@ def test_numpy(self):
         b = np.ones((2, 2)).astype("int64")
         self.assertEqual(a.numpy().dtype, b.dtype)
 
+    def test_redistribute(self):
+        # need to test with 1, 2, 3, and 4 dims
+        st = ht.zeros((50,), split=0)
+        if st.comm.size >= 3:
+            target_map = torch.zeros((st.comm.size, 1), dtype=torch.int)
+            target_map[1] = 30
+            target_map[2] = 20
+            st.redistribute_(target_map=target_map)
+            if st.comm.rank == 1:
+                self.assertEqual(st.lshape, (30,))
+            elif st.comm.rank == 2:
+                self.assertEqual(st.lshape, (20,))
+            else:
+                self.assertEqual(st.lshape, (0,))
+
+            st = ht.zeros((50, 50), split=1)
+            target_map = torch.zeros((st.comm.size, 2), dtype=torch.int)
+            target_map[0, 1] = 13
+            target_map[2, 1] = 50 - 13
+            st.redistribute_(target_map=target_map)
+            if st.comm.rank == 0:
+                self.assertEqual(st.lshape, (50, 13))
+            elif st.comm.rank == 2:
+                self.assertEqual(st.lshape, (50, 50 - 13))
+            else:
+                self.assertEqual(st.lshape, (50, 0))
+
+            st = ht.zeros((50, 81, 67), split=2)
+            target_map = torch.zeros((st.comm.size, 3), dtype=torch.int)
+            target_map[0, 2] = 67
+            st.redistribute_(target_map=target_map)
+            if st.comm.rank == 0:
+                self.assertEqual(st.lshape, (50, 81, 67))
+            else:
+                self.assertEqual(st.lshape, (50, 81, 0))
+
+            st = ht.zeros((8, 8, 8), split=None)
+            target_map = torch.zeros((st.comm.size, 3), dtype=torch.int)
+            # this will do nothing!
+            st.redistribute_(target_map=target_map)
+            self.assertTrue(st.lshape, st.gshape)
+
+            st = ht.zeros((50, 81, 67), split=0)
+            with self.assertRaises(ValueError):
+                target_map *= 0
+                st.redistribute_(target_map=target_map)
+
+            with self.assertRaises(TypeError):
+                st.redistribute_(target_map="sdfibn")
+            with self.assertRaises(TypeError):
+                st.redistribute_(lshape_map="sdfibn")
+            with self.assertRaises(ValueError):
+                st.redistribute_(lshape_map=torch.zeros(2))
+            with self.assertRaises(ValueError):
+                st.redistribute_(target_map=torch.zeros((2, 4)))
+
```

Comment:
OK I have a problem: when I run the unittests with n>=3 processes, I get this error from test_redistribute:

> Traceback (most recent call last):
>   File "/Users/c.comito/HAF/heat/heat/core/tests/test_dndarray.py", line 328, in test_redistribute
>     st.redistribute_(target_map=target_map)
>   File "/Users/c.comito/HAF/heat/heat/core/dndarray.py", line 2197, in redistribute_
>     self.redistribute_(lshape_map=lshape_map, target_map=target_map)
>   File "/Users/c.comito/HAF/heat/heat/core/dndarray.py", line 2197, in redistribute_
>     self.redistribute_(lshape_map=lshape_map, target_map=target_map)
>   File "/Users/c.comito/HAF/heat/heat/core/dndarray.py", line 2197, in redistribute_
>     self.redistribute_(lshape_map=lshape_map, target_map=target_map)
>   [Previous line repeated 968 more times]
>   File "/Users/c.comito/HAF/heat/heat/core/dndarray.py", line 2128, in redistribute_
>     if sum(target_map[..., self.split]) != self.shape[self.split]:
>   File "/anaconda3/envs/py37/lib/python3.7/site-packages/torch/tensor.py", line 411, in __iter__
>     return iter(imap(lambda i: self[i], range(self.size(0))))
> RecursionError: maximum recursion depth exceeded in comparison
>

Reply:
with the unit tests? mine runs fine. can you provide a failing example?

Reply:
Well, the test is the failing example. However, I only have this problem with python 3.7. I've just repeated the whole exercise on a new virtualenv with python 3.6, and the tests run fine for me as well.

Reply:
i just changed the sum to be an inline call. can you see if that changes it?